### ISSUE TITLE
Duplicate symbols libssl.a and libcrypto.a (3.0.8 regression)

### ISSUE THREAD

**Issue by nrgiii** [openssl/openssl#20238]
In 3.0.8 there are duplicate symbols in libssl.a and libcrypto.a that break the static link of our application.  Here's the linker output showing the duplicate symbols. Note v3.0.7 does not have this problem and links clean.

```
g++  -Wl,--whole-archive /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a  -Wl,--no-whole-archive main.o -o opensslmain -lpthread -lc -ldl

/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_new':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:59: multiple definition of `ossl_sa_new'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:59: first defined here
/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_free':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:111: multiple definition of `ossl_sa_free'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:111: first defined here
/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_free_leaves':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:119: multiple definition of `ossl_sa_free_leaves'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:119: first defined here
/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_doall':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:135: multiple definition of `ossl_sa_doall'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:135: first defined here
/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_doall_arg':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:146: multiple definition of `ossl_sa_doall_arg'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:146: first defined here
/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_num':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:152: multiple definition of `ossl_sa_num'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:152: first defined here
/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_get':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:157: multiple definition of `ossl_sa_get'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:157: first defined here
/usr/bin/ld: /home/normg/local/gs370/slow50/openssl/install50/lib/libcrypto.a(libcrypto-lib-sparse_array.o): in function `ossl_sa_set':
/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:180: multiple definition of `ossl_sa_set'; /home/normg/local/gs370/slow50/openssl/install50/lib/libssl.a(libssl-lib-sparse_array.o):/export/moop3/users/normg/gs370/slow50/openssl/../../externals/openssl/crypto/sparse_array.c:180: first defined here
collect2: error: ld returned 1 exit status
make: *** [Makefile:12: opensslmain] Error 1
```

**Comment by levitte** [comment]
Would you mind checking that #20240 fixes this issue?

**Comment by t8m** [comment]
Why do you use the `-Wl,--whole-archive`. I believe without this option the link would succeed.

**Comment by t8m** [comment]
Another workaround with the 3.0.8 possible is to use the `no-shared` option when building 3.0.8.

**Comment by nrgiii** [comment]
> Would you mind checking that #20240 fixes this issue?

Yes, #20240 fixes the issue.

**Comment by nrgiii** [comment]
Whoops, I spoke too soon. 
#20240 fixes the problem on Linux but not on Windows:


`
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_doall already defined in libcrypto-shlib-sparse_array.obj
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_doall_arg already defined in libcrypto-shlib-sparse_array.obj
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_free already defined in libcrypto-shlib-sparse_array.obj
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_free_leaves already defined in libcrypto-shlib-sparse_array.obj
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_get already defined in libcrypto-shlib-sparse_array.obj
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_new already defined in libcrypto-shlib-sparse_array.obj
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_num already defined in libcrypto-shlib-sparse_array.obj
libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_set already defined in libcrypto-shlib-sparse_array.obj
`

**Comment by t8m** [comment]
@nrgiii Did you do a clean build on Windows?

**Comment by nrgiii** [comment]
> @nrgiii Did you do a clean build on Windows?

Yes. I was wondering if Windows does not use build.obj? Any way I have a workaround on Windows where I modify the makefile and filter out the duplicate obj files.

**Comment by levitte** [comment]
I'm not sure what happens there.  The `libssl-shlib-` objects go strictly into the libssl DLL, and the `libcrypto-shlib-` objects go strictly into the libcrypto DLL, and those particular symbols are *not* (or shouldn't be) exported by the DLLs (or rather the corresponding import libraries).  So what's going on, I wonder?

@nrgii, what exactly did you do that lead up to this?  Was this when building OpenSSL itself, or was it when building your application?  How did you configure OpenSSL?

> ```
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_doall already defined in libcrypto-shlib-sparse_array.obj
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_doall_arg already defined in libcrypto-shlib-sparse_array.obj
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_free already defined in libcrypto-shlib-sparse_array.obj
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_free_leaves already defined in libcrypto-shlib-sparse_array.obj
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_get already defined in libcrypto-shlib-sparse_array.obj
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_new already defined in libcrypto-shlib-sparse_array.obj
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_num already defined in libcrypto-shlib-sparse_array.obj
> libssl-shlib-sparse_array.obj : error LNK2005: ossl_sa_set already defined in libcrypto-shlib-sparse_array.obj
> ```

**Comment by nrgiii** [comment]
> I'm not sure what happens there. The `libssl-shlib-` objects go strictly into the libssl DLL, and the `libcrypto-shlib-` objects go strictly into the libcrypto DLL, and those particular symbols are _not_ (or shouldn't be) exported by the DLLs (or rather the corresponding import libraries). So what's going on, I wonder?
> 
> @nrgii, what exactly did you do that lead up to this? Was this when building OpenSSL itself, or was it when building your application? How did you configure OpenSSL?

It's building our application, OpenSSL builds clean. We link the various .obj files used to build libcrypto and libssl and link them into a single DLL. The obj file names are extracted from the OpenSSL-generated makefile. A ilttle non-standard to be sure but we would still expect libssl.dll and libcrypto.dll to contain unique obj files without duplicates. Perhaps that assumption is not valid. We workaround this problem by filtering out the duplicate .obj files.

TLDR: OpenSSL builds clean and we are something weird after the build. If it's valid to have sparse_array.obj present in both libcrypto.dll and libssl.dll then this issue can be closed when PR #20240 is merged.

**Comment by levitte** [comment]
> It's building our application, OpenSSL builds clean. We link the various .obj files used to build libcrypto and libssl and link them into a single DLL. The obj file names are extracted from the OpenSSL-generated makefile.

Yup, I see the problem there, and since `crypto/sparse_array.c` is used *internally* in both DLLs, you get duplications when trying that...  mainly because the `SHARED_SOURCE` information in `build.info` doesn't get translated into the makefile, you're missing out on that piece of information.  Totally clear.
Side note: `configdata.pm` contains all that relevant information...  but I can understand that the information is structured in a way that's less easy to digest than it could have been.

> A little non-standard to be sure but we would still expect libssl.dll and libcrypto.dll to contain unique obj files without duplicates.

It's actually not very strange.  There are certain internal utility functionality that we don't want to expose to DLL users that is needed both in libcrypto and in libssl, and the technical solution becomes to include the same source file in both libraries, but only for the DLL form, since static libraries expose *everything* unconditionally, even things that are meant to be internal.

> Perhaps that assumption is not valid. We workaround this problem by filtering out the duplicate .obj files.

That's a reasonable workaround considering the problem space.  I'm glad you could do that.

> TLDR: OpenSSL builds clean and we are something weird after the build. If it's valid to have sparse_array.obj present in both libcrypto.dll and libssl.dll then this issue can be closed when PR #20240 is merged.

That's the conclusion I can see under current circumstances.

**Comment by levitte** [comment]
I merged #20240 just now

**Comment by wbl** [comment]
Should we add a CI check to prevent this in the future? I can work on one, but not sure what the best approach is.

**Comment by levitte** [comment]
> Should we add a CI check to prevent this in the future? I can work on one, but not sure what the best approach is.

It...  *can* be done by creating a small program that does nothing and that links with both static libraries wrapped with `-Wl,--whole-archive ` .. `-Wl,--no-whole-archive`, i.e. more or less what the original description demonstrates.  This should actually be perfectly possible to do entirely in a github action somewhere, limited to Ubuntu.

The extended option would be to have this done as part of our test suite (or the build tests), but that would have to be much more platform agnostic at the `build.info` level.  We actually have a few test programs that already link with both static libraries, which could serve as test builds for this sort of thing:

https://github.com/openssl/openssl/blob/cd870db16348d0d09cb05b7393cf9281509c7795/test/build.info#L600-L605

It's questionable if that's worth the entanglement...  initially, we can certainly go for the github action formula.

**Comment by mattcaswell** [comment]
> Should we add a CI check to prevent this in the future? 

I created issue #20304 for that based on @levitte's comment above. @wbl if you want to pick that up then please feel free.
